### PR TITLE
Clean-up and refactoring for 2.0

### LINKED ISSUE #673: Build pipeline workflow does not trigger
The build pipeline does not trigger.
According to @aepfli we introduced a bug in the branches section

> // EDIT: looks like we do have a bug in the yml and it should be actually `**`

### PR DESCRIPTION
**Pull request by Michael1993**
Closes nothing. A very self-indulgent PR/branch.

EDIT: Closes #673. Maybe?

Proposed commit message:

```
Refactor and document code (#672)

This commit simplifies some code with Java 11, adds some missing
documentation and fixes some code smells as marked by IntelliJ.

PR: #672
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [ ] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [ ] Site documentation in `.adoc` file references demo in `src/demo/java` instead of containing code blocks as text
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [ ] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [ ] The `package-info.java` contains information about the new extension

Code
* [ ] Code adheres to code style, naming conventions etc.
* [ ] Successful tests cover all changes
* [ ] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [ ] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional)

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -7,15 +7,14 @@ on:
...
-    branches: 'main'
+    branches: ['main']
     tags-ignore:
       - 'v**'
   # Each pull request is important to us, doesn't matter from which branch.
   # Furthermore, we do not want to build on just the default GitHub Action
   # events, we also want to react onto `labeled` events for our extended
   # build execution
   pull_request:
-    branches: '*'
```
Comment (Michael1993) [inline .github/workflows/build.yml:18]: Is this what you had in mind, @aepfli ?

File: src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java [hunk 2]
```diff
@@ -250,6 +259,16 @@ private static <A extends Annotation> List<A> findOnType(Class<?> element, Class
 				.collect(Collectors.toList());
 	}
 
+	/**
+	 * A helper utility method for {@link org.junitpioneer.jupiter.cartesian.CartesianTest} for finding
+	 * argument sources on parameters.
+	 *
+	 * @param testMethod the test method for which we want to find parameter arguments sources
+	 * @return a list of found arguments source annotations, including {@link CartesianArgumentsSource} annotations
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerAnnotationUtils.java:267]: I like to highlight if the `List` can be null or will at least return an not null, but empty list (e.g. via `Collections.emptyList()`

Comment fits many places, changed in this PR.

File: src/main/java/org/junitpioneer/internal/PioneerPreconditions.java [hunk 3]
```diff
@@ -60,10 +56,7 @@ public static String notBlank(String str, Supplier<String> messageSupplier) {
 	 * @return the supplied object
 	 */
 	public static <T> T notNull(T object, String message) {
-		if (object == null) {
-			throw new PreconditionViolationException(message);
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerPreconditions.java:64]: I think that this needs documentation update!

File: src/main/java/org/junitpioneer/internal/PioneerUtils.java [hunk 4]
```diff
@@ -165,14 +188,26 @@ public static List<List<?>> cartesianProduct(List<List<?>> lists) {
 		return resultLists;
 	}
 
+	/**
+	 * Static factory method for creating a Locale using {@link Locale.Builder}.
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerUtils.java:192]: missing `@param / @returns` javadoc

File: src/main/java/org/junitpioneer/internal/PioneerUtils.java [hunk 5]
```diff
@@ -165,14 +188,26 @@ public static List<List<?>> cartesianProduct(List<List<?>> lists) {
...
+	 * Required because Locale constructors are deprecated.
+	 */
 	public static Locale createLocale(String language, String country, String variant) {
 		return new Locale.Builder().setLanguage(language).setRegion(country).setVariant(variant).build();
 	}
 
+	/**
+	 * Static factory method for creating a Locale using {@link Locale.Builder}.
+	 * Required because Locale constructors are deprecated.
+	 */
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerUtils.java:202]: missing `@param / @returns` javadoc

File: src/main/java/org/junitpioneer/internal/PioneerUtils.java [hunk 6]
```diff
@@ -165,14 +188,26 @@ public static List<List<?>> cartesianProduct(List<List<?>> lists) {
...
 	public static Locale createLocale(String language, String country) {
 		return new Locale.Builder().setLanguage(language).setRegion(country).build();
 	}
 
+	/**
+	 * Static factory method for creating a Locale using {@link Locale.Builder}.
+	 * Required because Locale constructors are deprecated.
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/internal/PioneerUtils.java:209]: missing `@param / @returns` javadoc

File: src/main/java/org/junitpioneer/jupiter/EnvironmentVariableUtils.java [hunk 7]
```diff
@@ -68,8 +68,11 @@ private static void trySystemEnvClass(Consumer<Map<String, String>> consumer,
 	/*
 	 * Works on Windows
 	 */
+	@SuppressWarnings("Java9ReflectionClassVisibility")
 	private static void setInProcessEnvironmentClass(Consumer<Map<String, String>> consumer)
 			throws ReflectiveOperationException {
+		// The Java 9 access restriction problem is explicitly documented in our user guide
+		// for the EnvironmentVariable extension
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/EnvironmentVariableUtils.java:75]: maybe add a link here?

File: src/main/java/org/junitpioneer/jupiter/ReportEntryExtension.java [hunk 8]
```diff
@@ -135,7 +135,7 @@ private String parseVariables(String value, ExtensionContext context) {
 	}
 
 	private static boolean hasTestParameterVariables(String value) {
-		return value.matches(".*\\{[0-9]+}.*");
+		return value.matches(".*\\{\\d+}.*");
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/ReportEntryExtension.java:138]: I think `[0-9]` is easier to read / keep in mind than `\d`

**Review by Bukama — COMMENTED** [review]
Some small notes, most personal preferences. Most important ones most probably those about documentation update.

Can't say if changing the `! xx.isPresent()` into `isEmpty()` results in the same output - so I trust the build here :D

File: src/main/java/org/junitpioneer/internal/PioneerPreconditions.java (hunk 3 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/internal/PioneerPreconditions.java:64]: Sorry, I don't follow. Why do you think this needs a documentation update?

Comment (Bukama) [reply src/main/java/org/junitpioneer/internal/PioneerPreconditions.java:64]: I have in mind that this Exception is explicity named in the documentation of at least one of our extension. But I'm not at my private PC so I can't search in source code if I'm right.

Comment (Bukama) [reply src/main/java/org/junitpioneer/internal/PioneerPreconditions.java:64]: > I have in mind that this Exception is explicity named in the documentation of at least one of our extension. But I'm not at my private PC so I can't search in source code if I'm right.

After searching the files, it seem's I'm wrong and mixing it up with

> Specifying more than one kind of parameter source (i.e.: annotating your test parameters and the test method itself) does not work and will throw an `ExtensionConfigurationException`.

**Review by Bukama — APPROVED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (aepfli) [reply .github/workflows/build.yml:18]: yes

**Comment by Bukama** [comment]
Please provide a commit message, so we can merge this

**Pull request merged**